interpreter.go

```diff
@@ -146,33 +176,203 @@ type valueSimpleObjectField struct {
+	// has been inherited.  When evaluating a field from one of these super objects,
+	// we need to bind self to the concrete object (so self must point
+	// there) but uses of super should be resolved relative to the object whose
+	// field we are evaluating.  Thus, we keep a second field for that.  This is
+	// usually 0, unless we are evaluating a super object's field.
+	// TODO(sbarzowski) provide some examples
+	// TODO(sbarzowski) provide somewhere a complete explanation of the object model
+	// How deep in super we are. If we're executing
+	// TODO(sbarzowski) I was confused when I saw this name? Is it a standard term?
+	// If not maybe something along the lines of superDepth would be more obvious.
+	offset int
+
+	// Bindings introduced in this frame. The way previous bindings are treated
+	// depends on the type of a frame.
+	// If isCall == true then previous bindings are ignored (it's a clean
+	// environment with just the variables we have here).
+	// If isCall == false then if this frame doesn't contain a binding
+	// previous bindings will be used.
+	upValues bindingFrame
+}
+
+func makeEnvironment(upValues bindingFrame, self value, offset int) environment {
+	return environment{
+		upValues: upValues,
+		self:     self,
+		offset:   offset,
+	}
+}
+
+type thunk struct {
+	content value // nil if not filled
+	name    identifier
+	env     environment
+	body    astNode
+}
+
+func makeThunk(name identifier, env environment, body astNode) *thunk {
+	return &thunk{
+		name: name,
+		env:  env,
+		body: body,
+	}
+}
+
+func (t *thunk) fill(v value) {
+	t.content = v
+	// no point in keeping the environment - we already have the value
+	t.env = environment{}
+}
+
+func (t *thunk) filled() bool {
+	return t.content != nil
+}
+
+func (t *thunk) getValue(i *interpreter, loc *LocationRange) (value, error) {
+	if t.filled() {
+		// TODO(sbarzowski) what for? Stack trace?
+		// We're getting out of here anyway this seems useless
+		i.stack.newCall(loc, t, t.env)
+		return t.content, nil
+	}
+	return i.EvalInCleanEnv(loc, &t.env, t.body)
+}
+
 // The stack
 
-// TraceFrame is a single frame of the call stack.
+// TraceFrame is tracing information about a single frame of the call stack.
 type TraceFrame struct {
 	Loc  LocationRange
 	Name string
 }
 
 type callFrame struct {
-	isCall   bool
-	ast      astNode
-	location LocationRange
-	tailCall bool
-	thunks   []*thunk
-	context  value
-	self     value
-	offset   int
-	bindings bindingFrame
+	// True if it switches to a clean environment (function call or thunk)
+	// False otherwise, e.g. for local
+	// This makes callFrame a misnomer as it is technically not always a call...
+	isCall bool
+
+	// The code we were executing before.
+	// TODO(sbarzowski) what is it? Before? I guess it's the place from where it's
+	// called etc.
+	// TODO(sbarzowski) how could it be nil?
+	ast astNode
+
+	// The location of the code we were executing before.
```

Comment:
Even in the cases where the ast was null, I think there was still a useful LocationRange.